File: 백준/Gold/2118. 두 개의 탑/두 개의 탑.java
import java.util.*;
import java.io.*;

public class Main {

    static int N, total;
    static int[] dis;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        dis = new int[N + 1];
        for (int i = 0; i < N; i++) {
            int num = Integer.parseInt(br.readLine());
            total += num;
            dis[i] = num;
        }
        System.out.println(solution());
    }

    public static int solution() {
        int left = 0, right = 0, max = -1;
        int now = dis[0];
        while(left <= right && right < N){
            int diff = Math.min(now, total - now);
            max = Math.max(max, diff);

            if(now == diff){
                right ++;
                now += dis[right];
            }else{
                now -= dis[left++];
            }
        }

        return max;
    }
}

File: 백준/Gold/2118. 두 개의 탑/두 개의 탑.py
import sys
input = sys.stdin.readline

num = int(input())
arr = [int(input().strip()) for _ in range(num)]

r, cur, ans = 0, 0, 0
total = sum(arr)
for l in range(num):
    while(cur < total // 2):
        cur += arr[r % num]
        r += 1

    ans = max(ans, min(cur, total - cur))
    cur -= arr[l]

print(ans)